File: src/2024/历年真题/2024-国赛全真模拟测试卷/下/03.py
# https://www.lanqiao.cn/problems/17133/learning/


def quickPower(a, b, mod):
    res = 1
    base = a
    while b > 0:
        if b & 1:
            res = (res*base) % mod
        base = (base*base) % mod
        b >>= 1
    return res


MOD = 2146516019
ans = 0  # 0 异或任何数等于该数本身
for i in range(1, 1+233333333):
    ans ^= quickPower(i, MOD-2, MOD)

print(ans)
# print(1307261675)

File: src/2024/历年真题/2024-国赛全真模拟测试卷/下/03.cpp
#include <iostream>
using ll = long long;

ll quickPower(ll a, ll b, ll mod)
{
    ll res = 1, base = a;
    while (b > 0)
    {
        if (b & 1)
        {
            res = (res * base) % mod;
        }
        base = (base * base) % mod;
        b >>= 1;
    }
    return res;
}

int main()
{
    ll MOD = 2146516019;
    ll ans = 0;

    for (ll i = 1; i <= 233333333; ++i)
    {
        ans ^= quickPower(i, MOD - 2, MOD);
    }

    std::cout << ans << '\n';
    // std::cout << 1307261675 << '\n';

    return 0;
}
